Initialize random number generator in early boot

raghav-g13 opened the pull request:
The random number generator (most commonly `/dev/urandom`) is not initialized (and thus blocks on calls) until it hits a minimum entropy threshold, currently 256 (bits of entropy). 

This manifests as applications hanging unexplicably and then running as expected after a long time period. Hitting the minimum entropy threshold in a deterministic simulated environment requires lots of cycles and IO ops. We bypass the issue by manually overriding the entropy value during early boot. 

The override uses two Kconfig options:
- OVERRIDE_RNG_ENTROPY to enable the override
- OVERRIDE_RNG_ENTROPY_VALUE for the override value
   * Depends on OVERRIDE_RNG_ENTROPY

raghav-g13 committed:
Update random.c to initialize rng in early boot

abejgonzalez commented on drivers/char/random.c:
```diff
@@ -871,8 +871,18 @@ void __init random_init_early(const char *command_line)
-		_credit_init_bits(arch_bits);
+	else if (trust_cpu) {
+
+		// _credit_init_bits(arch_bits);
```
What happens if you just do `random.trust_cpu=off` in the kernel cmdline?

jerryz123 replied:
Isn't the default going to be `random.trust_cpu=off`? Is the intention for this patch to be used with `random.trust_cpu=on`?

raghav-g13 replied:
this change will get bypassed, so easy to switch out!

raghav-g13 replied:
yes, the intention is to use it with `random.trust_cpu=on`

will implement that in marshal (firechip br-base linux-config)

abejgonzalez commented on drivers/char/random.c:
```diff
@@ -871,8 +871,18 @@ void __init random_init_early(const char *command_line)
+   		 * sets entropy to 256 to initialize random number geenrator
+      		 * in early boot.
+		 */
+		size_t artifical_arch_bits = 256;
```
Why did you choose 256? What is the threshold?

raghav-g13 replied:
256 is the threshold, lemme comment that

raghav-g13 committed:
Update random.c

raghav-g13 committed:
Update random.c

Switch to a setup with two Kconfig options

OVERRIDE_RNG_ENTROPY to enable
OVERRIDE_RNG_ENTROPY_VALUE for the override value

raghav-g13 committed:
Update Kconfig

Switch to a setup with two Kconfig options

OVERRIDE_RNG_ENTROPY to enable
OVERRIDE_RNG_ENTROPY_VALUE for the override value

jerryz123 commented:
When you update the fire marshal configs, can you remember to update them for all the targets, not just firechip

abejgonzalez reviewed (approved):
LGTM
